### PR TITLE
t/291: ToolbarView#fillFromConfig should warn when the factory does not provide a component

### LINKED ISSUE #291: The ToolbarView should warn when the factory does not provide a component
A follow-up of https://github.com/ckeditor/ckeditor5/issues/526.

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Other: ToolbarView#fillFromConfig should warn when the factory does not provide a component. Closes #291. Closes https://github.com/ckeditor/ckeditor5/issues/526.

---

### Additional information

* Could close https://github.com/ckeditor/ckeditor5/issues/526.

### CODE REVIEW

File: src/toolbar/toolbarview.js [hunk 1]
```diff
@@ -126,9 +127,29 @@ export default class ToolbarView extends View {
...
+				 * This warning usually shows up when the plugin that is supposed to register the toolbar
+				 * component in the factory has not been loaded or there's a typo in the configuration
+				 * of the toolbar.
+				 *
+				 * @error toolbarview-missing-component
+				 * @param {String} name The name of the component.
+				 * @param {module:ui/componentfactory~ComponentFactory} factory The factory that is missing the component.
+				 */
+				log.warn(
+					'toolbarview-missing-component: There is no such component in the factory.',
```
Comment (Reinmar) [inline src/toolbar/toolbarview.js]: This was supposed to be a more meaningful error, but it's still all about esoteric components and factories. I think you should read https://medium.com/@thomasfuchs/how-to-write-an-error-message-883718173322 ;)

File: src/toolbar/toolbarview.js [hunk 2]
```diff
@@ -126,9 +127,29 @@ export default class ToolbarView extends View {
...
-			const component = name == '|' ? new ToolbarSeparatorView() : factory.create( name );
-
-			this.items.add( component );
+			if ( name == '|' ) {
+				this.items.add( new ToolbarSeparatorView() );
+			} else if ( factory.has( name ) ) {
+				this.items.add( factory.create( name ) );
+			} else {
+				/**
+				 * There was a problem with expanding the toolbar configuration into toolbar items.
```
Comment (Reinmar) [inline src/toolbar/toolbarview.js]: Same here as below – "expanding" – really? There should be a registry of forbidden words. Here, try not to use "expanding" and "factory" and "components" :P

**Comment by Reinmar** [comment]
R- because of the wording (which is the essential part o this ticket).

**Comment by oleq** [comment]
Hopefully, this will look more understandable now.

**Comment by Reinmar** [comment]
Since there's only one factory of UI items in the editor I'll remove it from the error details. This will make reading the error easier since there will be only the critical information.

**Comment by oleq** [comment]
> Since there's only one factory of UI items in the editor I'll remove it from the error details. This will make reading the error easier since there will be only the critical information.

Don't forget that ckeditor5-ui is not only about CKEditor, it's a UI framework and it could be used anywhere, each time with a different factory.

**Comment by Reinmar** [comment]
> Don't forget that ckeditor5-ui is not only about CKEditor, it's a UI framework and it could be used anywhere, each time with a different factory.

Sure. And this will happen how many times? ;)

Besides, in such special use cases, the developer will be more aware of what they are doing anyway.

**Pull request merged**